### PR TITLE
[crypto/ec] don't assume points are of order group->order

### PR DESCRIPTION
**Pull request by bbbrumley**
Part of #6009 assumes points have order `group->order`. That's not always the case.

This PR only assumes points have order dividing the cardinality of the curve (group). That's an abstract algebra guarantee, as long as the point lies on the curve.

Negative testing for co-factor ECDH should've caught this. But when I added #763 (later converted to evptests by @romen in #3218) the KATs were copy/paste from NIST. And there aren't any negative tests. Really disappointing from NIST, since that's one of the main points of co-factor ECDH.

So I added one negative test that would've caught it. Follow up: more negative testing for "weird" points.

Tagging @mattcaswell

### CODE REVIEW

**Comment by kroeckx** [comment]
I think negative testing is just as important as positive testing in most cases.

Anyway, thank you and your team for all the work you're putting into test.

File: test/recipes/30-test_evp_data/evppkey.txt [hunk 1]
```diff
@@ -17257,6 +17257,35 @@ PeerKey=KAS-ECC-CDH_B-571_C24-Peer-PUBLIC
 Ctrl=ecdh_cofactor_mode:1
 SharedSecret=02da266a269bdc8d8b2a0c6bb5762f102fc801c8d5394a9271539136bd81d4b69cfbb7525cd0a983fb7f7e9deec583b8f8e574c6184b2d79831ec770649e484dc006fa35b0bffd0b
 
+# for cofactor-order points, CECDH should fail. Test that.
```
Comment (bernd-edlinger) [inline test/recipes/30-test_evp_data/evppkey.txt]: did you mean ECDH, or what is CECDH ?

Comment (bbbrumley) [reply test/recipes/30-test_evp_data/evppkey.txt]: > did you mean ECDH, or what is CECDH ?

I meant co-factor ECDH, but perhaps I just made up the CECDH acronym :) Technically it's ECC CDH. I'll fix it. (The `Ctrl` entry toggles this ECDH variant.)

**Comment by levitte** [comment]
Should this be backported to 1.1.0?

**Comment by bbbrumley** [comment]
> Approval for 1.1.0 as well, if we decide to backport

Yes def needs a backport since this is a regression caused by #6009. @sohhas or @romen, would you mind?

**Comment by levitte** [comment]
> Yes def needs a backport since this is a regression caused by #6009

That's good enough reason for me

**Comment by bernd-edlinger** [comment]
I would not mind if the tests are not back-ported.

**Comment by romen** [comment]
I'll create the backport PR

**Comment by romen** [comment]
@levitte so #6549 (the 1.1.0 backport of this) got merged already, this PR seems to be ready too, it's just missing an assignee and should just lose the 1.1.0 flag!

**Comment by levitte** [comment]
Ok, nothing changed since @paulidale's approval, so I'm merging

**Comment by levitte** [comment]
Merged.

a766aab93a282774e63ba918d0bb1c6680a5f292 [crypto/ec] don't assume points are of order group->order

**Pull request closed without merging**